### DISCUSSION TITLE
[Question] Example of enumeration with MultiStateValueDiscreteTypeNode in Server SDK

### DISCUSSION THREAD

**Discussion by devaskim** [eclipse-milo/milo#1328] [General]
Could someone explain me when to call [MultiStateValueDiscreteTypeNode#setValueAsText()](https://github.com/eclipse/milo/blob/master/opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/model/nodes/variables/MultiStateValueDiscreteTypeNode.java#L71) ?

I created instance of MultiStateValueDiscreteTypeNode, [set array of EnumValueType](https://github.com/eclipse/milo/blob/master/opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/model/nodes/variables/MultiStateValueDiscreteTypeNode.java#L54) , but after connecting to server I can set only integer value. My expectation was to see dropdown list with string values in it.

**Comment by kevinherron** [comment]
See the description of this property at https://reference.opcfoundation.org/Core/Part3/v105/docs/5.6

It does not replace the value. It’s only used by a client that is aware of it and chooses to do something with it. 

> This optional Property provides the localized text representation of the value. It can be used by Clients only interested in displaying the text to subscribe to the Property instead of the value attribute.

**Reply by kevinherron** [reply]
Also, a Client that wanted to display a dropdown of all values would need to use the `EnumStrings` or `EnumValues` property on the enum's DataType Node, not this one.

**Reply by devaskim** [reply]
> Also, a Client that wanted to display a dropdown of all values would need to use the `EnumStrings` or `EnumValues` property on the enum's DataType Node, not this one.

Maybe you know if UaExpert client supports such dropdowns? Because I as wrote in the issue description, I had set  `EnumValues` property, but it has not affect on a client side.

![Скриншот 25-09-2024 122556](https://github.com/user-attachments/assets/bb7a8df8-7f50-4e89-b876-fbb7137cbaf5)

**Reply by kevinherron** [reply]
UaExpert does support it, but you have to configure your VariableNode so that it has a DataType that is a subtype of Enumeration. This DataType Node is where the enum strings are defined.

You currently have the DataType configured as Int32.

**Reply by devaskim** [reply]
>  This DataType Node is where the enum strings are defined.

Just to ensure, `ENUM_STRINGS` is about `MultiStateDiscreteType`, but I try to use `MultiStateValueDiscreteTypeNode` with type `MultiStateValueDiscreteType` where only `ENUM_VALUES` and `VALUE_AS_TEXT`. So to have a dropdown with strings I need `MultiStateDiscreteType` or `MultiStateValueDiscreteType` ?

With the latter UaExpert shows the following
![Скриншот 26-09-2024 190113](https://github.com/user-attachments/assets/a3e34a7f-eadd-4c45-be7f-649293bd0568)

**Reply by kevinherron** [reply]
`EnumValueType` is not an enumeration. You are not understanding that the DataType of your Value needs to be a subclass of the Enumeration type, e.g. `ApplicationType`.

```
MultiStateDiscreteTypeNode node = new MultiStateDiscreteTypeNode(
    getNodeContext(),
    newNodeId("HelloWorld/MyMultiStateDiscreteVariable"),
    newQualifiedName("MyMultiStateDiscreteVariable"),
    LocalizedText.english("MyMultiStateDiscreteVariable"),
    LocalizedText.NULL_VALUE,
    uint(0),
    uint(0),
    null,
    null,
    null,
    new DataValue(Variant.ofInt32(0)),
    NodeIds.ApplicationType,
    -1,
    null,
    AccessLevel.toValue(AccessLevel.READ_WRITE),
    AccessLevel.toValue(AccessLevel.READ_WRITE),
    100.0,
    false,
    null
);

node.addReference(new Reference(
    node.getNodeId(),
    NodeIds.HasTypeDefinition,
    NodeIds.MultiStateDiscreteType.expanded(),
    true
));
node.addReference(new Reference(
    node.getNodeId(),
    NodeIds.Organizes,
    folderNodeId.expanded(),
    false
));
getNodeManager().addNode(node);
```

In this case UaExpert will see that the DataType of the Value is `ApplicationType` and know how to get the enum strings.

`MultiStateDiscreteType` is a _VariableType_, it's not enough. You have to actually create a DataType in your server that is a subclass of the Enumeration type and configure your VariableNode instance (whether it's a MultiStateDiscreteType or not) to use that datatype.

<img width="660" alt="Screenshot 2024-09-26 at 07 24 13" src="https://github.com/user-attachments/assets/9a481e34-b394-4449-a371-88c907508315">
<img width="661" alt="Screenshot 2024-09-26 at 07 24 19" src="https://github.com/user-attachments/assets/067308fe-190c-4af1-8914-655ebf4c5bb6">

**Reply by kevinherron** [reply]
Look at ExampleNamespace, which shows the how to create a custom enum type, register it, and adds a Node that uses it.

**Reply by devaskim** [reply]
> Look at ExampleNamespace, which shows the how to create a custom enum type, register it, and adds a Node that uses it.

I saw examples, but such custom enum should be defined on compilation stage, i.e. you cannot create it dynamically. My case requires creating enums having only Java's `Map<Long, String>`-s with various values in it. So I thought OPC UA provides some data type for such scenario, where I could just put index-to-description mapping on a server side, and client will show desired dropdown and limits possible input values.

**Reply by kevinherron** [reply]
Everything the example does is still applicable even if you don't actually create a Java `enum` class. You would just set Int32 values on the Node. But all the rest about creating a DataType and making sure it's registered with the server must be done and can be done without the "static" part.

**Reply by devaskim** [reply]
Many thanks, Kevin, for you detailed explanation. I will take a look again at examples and try to code
